src/core/collision.cpp

```diff
@@ -654,110 +580,165 @@ void three_particle_binding_domain_decomposition()
+	   place_vs_and_relate_to_particle(current_vs_pid,pos1,c.pp1,initial_pos);
+     current_vs_pid++;
+	   place_vs_and_relate_to_particle(current_vs_pid,pos2,c.pp2,initial_pos);
+     current_vs_pid++;
+     bind_at_poc_create_bond_between_vs(current_vs_pid,c);
+   }
+   else
+   { // Just update the books
+        added_particle(current_vs_pid);
+        current_vs_pid++;
+        
+        added_particle(current_vs_pid);
+        current_vs_pid++;
+   }
 	}
-	// The virtual site related to p2 is needed independently on which of the vs-related modes is active
-        place_vs_and_relate_to_particle(collision_queue[i].point_of_collision,collision_queue[i].pp2);
-  
-	// If we are in the two vs mode, we need a bond between the virtual sites
-	if (collision_params.mode & COLLISION_MODE_VS)
-	{
-          bind_at_poc_create_bond_between_vs(i);
-        }
-	
-	// If we are in the "glue to surface mode", we need a bond between p1 and the vs
-	if (collision_params.mode & COLLISION_MODE_GLUE_TO_SURF)
-	{
-           glue_to_surface_bind_vs_to_pp1(i);
-        }
+	if (collision_params.mode & COLLISION_MODE_GLUE_TO_SURF) {
+      double pos[3];
+      const int pid=glue_to_surface_calc_vs_pos(p1,p2,pos);
+	    
+      // Change type of partilce being attached, to make it inert
+      if (p1->p.type==collision_params.part_type_to_be_glued) {
+         p1->p.type=collision_params.part_type_after_glueing;
+      }
+      if (p2->p.type==collision_params.part_type_to_be_glued) {
+         p2->p.type=collision_params.part_type_after_glueing;
+      }
+      
+      // Vs placement happens on the node that has p1
+      if (!p1->l.ghost) {
+        place_vs_and_relate_to_particle(current_vs_pid,pos,pid,initial_pos);
+        current_vs_pid++;
+      } 
+      else 
+      { // Just update the books
+        added_particle(current_vs_pid);
+        current_vs_pid++;
+      }  
+      glue_to_surface_bind_part_to_vs(p1,p2,current_vs_pid,c);
+      
+    }
       } // Loop over all collisions in the queue
+
+      // If any node had a collision, all nodes need to do on_particle_change
+      // and resort
+
+      if (gathered_queue.size()>0) {
+        on_particle_change();
+        announce_resort_particles();
+      }
+     } // total_collision>0
     } // are we in one of the vs_based methods
 #endif //defined VIRTUAL_SITES_RELATIVE
   
 
+  
   // three-particle-binding part
-
-
   if (collision_params.mode & (COLLISION_MODE_BIND_THREE_PARTICLES)) {
-  std::vector<int> counts(n_nodes);
-  gather_collision_queue(counts.data());
+  auto gathered_queue = gather_global_collision_queue();
 
-    if (counts[this_node]>0) {
 
       // If we don't have domain decomposition, we need to do a full sweep over all
       // particles in the system. (slow)
       if (cell_structure.type!=CELL_STRUCTURE_DOMDEC) {
-        three_particle_binding_full_search();
+        three_particle_binding_full_search(gathered_queue);
```

Comment:
If the full search only works for 1 process and aborts if it is called on several, why do we need to gather the collision queue at all?